File: java-workshop/part-3-mocks-clean/src/test/java/service/DiscountCodeServiceTest.java
package service;

import domain.DiscountCode;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import repository.DiscountCodeRepository;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class DiscountCodeServiceTest {

    @Mock
    private DiscountCodeRepository discountCodeRepository;

    @InjectMocks
    private DiscountCodeService discountCodeService;

    @Test
    void checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed() {
        when(discountCodeRepository.checkDiscountCode("CUST-001", DiscountCode.SUMMER20)).thenReturn(true);

        assertTrue(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    @Test
    void checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed() {
        when(discountCodeRepository.checkDiscountCode("CUST-001", DiscountCode.SUMMER20)).thenReturn(false);

        assertFalse(discountCodeService.checkDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    @Test
    void markAsUsed_delegatesToRepository() {
        discountCodeService.markAsUsed("CUST-001", DiscountCode.SUMMER20);

        verify(discountCodeRepository).markAsUsed("CUST-001", DiscountCode.SUMMER20);
    }
}

File: csharp-workshop/Part3.MocksClean.Tests/Service/DiscountCodeServiceTest.cs
using NSubstitute;
using Part3.MocksClean.Domain;
using Part3.MocksClean.Repository;
using Part3.MocksClean.Service;
using Xunit;

namespace Part3.MocksClean.Tests.Service;

public class DiscountCodeServiceTest
{
    private readonly IDiscountCodeRepository _discountCodeRepository = Substitute.For<IDiscountCodeRepository>();
    private readonly DiscountCodeService _discountCodeService;

    public DiscountCodeServiceTest()
    {
        _discountCodeService = new DiscountCodeService(_discountCodeRepository);
    }

    [Fact]
    public void P3_CheckDiscountCode_ReturnsTrue_WhenCodeHasAlreadyBeenUsed()
    {
        _discountCodeRepository.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20).Returns(true);

        Assert.True(_discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P3_CheckDiscountCode_ReturnsFalse_WhenCodeHasNotBeenUsed()
    {
        _discountCodeRepository.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20).Returns(false);

        Assert.False(_discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P3_MarkAsUsed_DelegatesToRepository()
    {
        _discountCodeService.MarkAsUsed("CUST-001", DiscountCode.SUMMER20);

        _discountCodeRepository.Received(1).MarkAsUsed("CUST-001", DiscountCode.SUMMER20);
    }
}
